x/mongo/driver/batch_cursor.go

```diff
@@ -364,17 +364,10 @@ func (bc *BatchCursor) getMore(ctx context.Context) {
 		}
 	}
 
-	// If we're in load balanced mode and the pinned connection encounters a network error, it should be unpinned. In
-	// this case, we call Expire on the connection to ensure it's closed and the cursor will be reaped by the server.
+	// If we're in load balanced mode and the pinned connection encounters a network error, we should not use it for
+	// future commands. Per the spec, the connection will not be unpinned until the cursor is actually closed, but
+	// we set the cursor ID to 0 to ensure the Close() call will not execute a killCursors command.
 	if driverErr, ok := bc.err.(Error); ok && driverErr.NetworkError() && bc.connection != nil {
-		err := bc.connection.Expire()
-		if err != nil && bc.err == nil {
-			bc.err = err
-		}
-
-		// Also unset the connection and set the cursor ID to 0 because the cursor is no longer valid, so we shouldn't
-		// send any more commands for it.
-		bc.connection = nil
```

Comment:
The `Connection#pin` function only updates the pool statistics for the very first reference (see [this conditional](https://github.com/mongodb/mongo-go-driver/blob/ad923e2619e149cea42ad905137020b8d05f019f/x/mongo/driver/topology/connection.go#L775)) to avoid double-counting in the case of one connection being pinned by both a cursor and a transaction. In the case you described, step (2)  would call `Connection#PinToTransaction`, which would update the pool by calling `pool#pinConnectionToTransaction`. In step (3), we'd call `Connection#PinToCursor`, which would increment the reference count of the `Connection` object but would not call `pool#pinConnectionToCursor`. Once the cursor is drained/closed, we'd call `Connection#UnpinFromCursor`, which would decrement the ref count, and `Connection#Close`, which would be a no-op because the ref count is still non-zero.

This [unit test](https://github.com/mongodb/mongo-go-driver/blob/ad923e2619e149cea42ad905137020b8d05f019f/x/mongo/driver/topology/connection_test.go#L900) for the `Connection` type performs a similar set of steps, so maybe the assertions there can help clarify what happens too.

Reply:
Ah, thank you for linking to that unit test, that does clarify the expected behavior. I'm now remembering the necessity to avoid double counting. In the case I described, only the transaction pinning would be counted. Sorry for rehashing this – I should have remembered that from the review for GODRIVER-1901.